### PR TITLE
[FLINK-18128][FLINK-19223][connectors] Fixes to Connector Base Availability logic

### PR DESCRIPTION
**Pull request by StephanEwen**
## What is the purpose of the change

This PR fixes bugs in the connector base (`SourceReaderBase` and `SplitFetcher`) availability logic.
This could previously lead to conditions where the reader was done (all `SplitFetchers` idle) but this was never recognized.
The result was deadlocking jobs.

## Brief change log

The fix contains two steps, the problem and rational is described in the Jira issues:
  - https://issues.apache.org/jira/browse/FLINK-18128
  - https://issues.apache.org/jira/browse/FLINK-19223

## Verifying this change

This change is partially covered by existing tests and also  adds tests. See changes in `SplitFetcherTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **yes**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherTest.java [hunk 1]
```diff
@@ -29,25 +35,94 @@
...
+		final SplitFetcher<Object, TestingSourceSplit> fetcher = createFetcherWithSplit(
+			"test-split",
+			queue,
+			new TestingSplitReader<>(finishedSplitFetch("test-split")));
+
+		fetcher.runOnce();
+
+		assertTrue(fetcher.assignedSplits().isEmpty());
+		assertTrue(fetcher.isIdle());
+		assertTrue(queue.getAvailabilityFuture().isDone());
```
Comment (becketqin) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherTest.java:95]: This tests that the queue returns an available future after the fetcher goes to idle. Should we get a future before the fetcher goes to idle and assert that future is done here?

File: tools/ci/log4j.properties [hunk 2]
```diff
@@ -73,3 +73,8 @@ logger.consumer.level = OFF
 logger.queryablestate.name = org.apache.flink.queryablestate
 logger.queryablestate.level = TRACE
 logger.queryablestate.appenderRef.out.ref = ConsoleAppender
+
+# Temporarily enable TRACE for SourceReaderBase to debug a deadlock
+logger.sourcereader.name = org.apache.flink.connector.base.source.reader
+logger.sourcereader.level = TRACE
```
Comment (becketqin) [inline tools/ci/log4j.properties:79]: Are we going to keep this trace logging?

**Review by becketqin — APPROVED** [review]
@StephanEwen Thanks for the patch and fixing the deadlock. The Patch looks good to me with just some minor comments.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/synchronization/FutureCompletingBlockingQueue.java [hunk 3]
```diff
@@ -275,4 +340,22 @@ private void setWakeUp(boolean value) {
 			wakeUp = value;
 		}
 	}
+
+	// ------------------------------------------------------------------------
+	//  utilities
+	// ------------------------------------------------------------------------
+
+	@SuppressWarnings("unchecked")
+	private static CompletableFuture<Void> getAvailableFuture() {
```
[the remark this answers is not in the document]
Comment (StephanEwen) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/synchronization/FutureCompletingBlockingQueue.java:349]: I would like to keep this. It does not add a hard dependency, it is a best-effort optimization.

This future will in the end be passed all the way to the main mailbox. There it helps to short circuit certain conditions, eliding volatile memory accesses. That is a good optimization, in my opinion. Any volatile memory access we can eliminate on the per-record paths tend to be worth it. They make the single thread faster but also reduce stress on cache coherency protocol, etc.

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherTest.java (hunk 1 above)
Comment (StephanEwen) [reply flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherTest.java:95]: Good idea, let's have both tests.

File: tools/ci/log4j.properties (hunk 2 above)
Comment (StephanEwen) [reply tools/ci/log4j.properties:79]: We could (it does not affect any other tests) but I was planning to remove this once we are confident with the implementation.

Comment (StephanEwen) [reply tools/ci/log4j.properties:79]: Let me remove it during merging. We can always add it back if needed.

**Pull request closed without merging**